src/mame/drivers/krz2000.cpp

```diff
@@ -52,63 +65,315 @@ class k2000_state : public driver_device
+	// map(0x300000, 0x3fffff).rom().region("pcm", 0x100000); // Block 3: Sample ROM
+	// map(0x400000, 0x4fffff).rom().region("pcm", 0x200000); // Block 4: Sample ROM
+	// Writes to 0x4FFD82, 0x4FFD88, 0x4FFE24
+
+	// Hypothesis: Calvin (or Janis) is mapped at 0x500000
+	//  Writes to 0x500402, 0x500418, 0x50041A early in the boot process
+	//  Initialization then writes zeros in a descending ladder:
+	//  0x500400, 0x500402
+	//  0x5003FC, 0x5003FE, ...
+	//  down to 0x500004 (then 0x500006)
+
+	// Calvin memory?
+	map(0x500000, 0x5003FF).rw(FUNC(k2000_state::calvin_read), FUNC(k2000_state::calvin_write));
+	// Calvin control registers?
+	map(0x500400, 0x5004FF).rw(FUNC(k2000_state::calvin_read), FUNC(k2000_state::calvin_write));
+	// Reads and writes to 0x580000
+
+	// Hobbes (it seems) is connected in the 0x600000 range
+	// The second Hobbes chip (?) seems connected at 0x680000
+	// Initialization seems to:
+	//  - write to 0x600300, followed by 0x600xxx where xxx is in 0x400 through 0x5FF
+	//  - write to 0x6?0002, 0x6?0006, 0x6?0004, through 0x6?000F4
+	//  - write to 0x600004, 0x600014, 0x600024... through 0x6002F4
+	//    (with a repeating pattern of [0x02, 0x00, 0x26, 0x24, 0x4A, 0x48])
+	//  - write to 0x680004, 0x680014, 0x680024... through 0x6802F4
+	//    (with a repeating pattern of [0x80, 0x02, 0xA4, 0x26, 0xC6, 0x4A])
+	// Each Hobbes chip seems to have two control registers, mapped at 0x600300/0x680300 and 0x600400/0x680400
+
+	// Hobbes memory?
+	map(0x600000, 0x6002FF).rw(FUNC(k2000_state::hobbes0_read), FUNC(k2000_state::hobbes0_write));
+	// Hobbes control registers?
+	map(0x600300, 0x6005FF).rw(FUNC(k2000_state::hobbes0_read), FUNC(k2000_state::hobbes0_write));
+
+	// Hobbes memory?
+	map(0x680000, 0x6802FF).rw(FUNC(k2000_state::hobbes1_read), FUNC(k2000_state::hobbes1_write));
+	// Hobbes control registers?
+	map(0x680300, 0x6805FF).rw(FUNC(k2000_state::hobbes1_read), FUNC(k2000_state::hobbes1_write));
+
+	// SROM - Setup ROM?
+	uint32 srom_base = 0x280000;
+	map(srom_base, srom_base + 0x40000 - 1).rom().region("maincpu", 0x100000);
+
+	// Calvin chip has a 20-bit address bus with A0 tied low and A12-15 tied high, giving it access to 0x?0??[0-E]
+	// Later models' Janis chips use 24-bit address buses (or at least they're all connected in the schematic)
+	// Hobbes has a 10-bit address bus with A0 tied low, giving access to 0x0 through 0x3FF
+
+	// Block 7: LCD control and bank switching
 	map(0x700000, 0x700003).rw("lcd", FUNC(lm24014h_device::read), FUNC(lm24014h_device::write)).umask16(0x00ff);
-	map(0x7e0000, 0x7e0001).w(FUNC(k2000_state::ctrl_w));
+
+	// Reads from 0x740000 (floppy?)
+	// Writes to 0x760000
+	// Writes to 0x7C0000
+
+	// Single byte reads seem to happen at 0x740000, which seems to be FDCCTLCSb?
+	// map(0x740000, 0x74000F).
+	map(0x7e0000, 0x7e0001).w(FUNC(k2000_state::swap_bit_ctrl_w));
+
+	// Some sort of hardware mapping - initial firmware boot expects a 4 at $78000B
+	map(0x780000, 0x78FFFF).ram();
```

Comment:
address_map mixes up UPPERCASE and lowercase in mapping. Please use latter only.